benchmark/matmul.cpp

```diff
@@ -377,3 +440,329 @@ ForAllLayouts(NvFuserScheduler_4warp4stage_test);
+  run(TT, MatmulLayout::TT, splitk_factor);     \
+  run(TN, MatmulLayout::TN, splitk_factor);     \
+  run(NT, MatmulLayout::NT, splitk_factor);     \
+  run(NN, MatmulLayout::NN, splitk_factor);
+
+// Automatically determine splitk_factor to fill one wave
+
+#define AutoSplitK(layout_label)                              \
+  BENCHMARK_CAPTURE(                                          \
+      NvFuserScheduler_Matmul_4warp3stage,                    \
+      nvfuser_108block_auto_splitk_##layout_label,            \
+      MatmulLayout::layout_label,                             \
+      -1)                                                     \
+      ->SplitKMatmulA100Shapes->Unit(benchmark::kMicrosecond) \
+      ->UseManualTime();
+
+AutoSplitK(TT);
+AutoSplitK(TN);
+AutoSplitK(NT);
+AutoSplitK(NN);
+
+SplitKForAllLayouts(NvFuserScheduler_4warp3stage_splitk_test, 1);
+SplitKForAllLayouts(NvFuserScheduler_4warp3stage_splitk_test, 2);
+SplitKForAllLayouts(NvFuserScheduler_4warp3stage_splitk_test, 4);
+SplitKForAllLayouts(NvFuserScheduler_4warp3stage_splitk_test, 8);
+SplitKForAllLayouts(NvFuserScheduler_4warp3stage_splitk_test, 16);
+SplitKForAllLayouts(NvFuserScheduler_4warp3stage_splitk_test, 32);
+
+SplitKForAllLayouts(NvFuserScheduler_8warp4stage_splitk_test, 1);
+SplitKForAllLayouts(NvFuserScheduler_8warp4stage_splitk_test, 2);
+SplitKForAllLayouts(NvFuserScheduler_8warp4stage_splitk_test, 4);
+SplitKForAllLayouts(NvFuserScheduler_8warp4stage_splitk_test, 8);
+SplitKForAllLayouts(NvFuserScheduler_8warp4stage_splitk_test, 16);
+SplitKForAllLayouts(NvFuserScheduler_8warp4stage_splitk_test, 32);
+
+// This performs the splitk matmul WITHOUT any outer reduction, which is useful
+// for comparing against the first kernel in Cutlass's two-kernel split-K.
+static void SingleMatmulPartitionedK(
+    benchmark::State& benchmark_state,
+    MatmulLayout layout,
+    MatmulParams params,
+    int64_t splitk_factor) {
+  int64_t M = benchmark_state.range(0);
+  int64_t N = benchmark_state.range(1);
+  int64_t K = benchmark_state.range(2);
+
+  NVF_CHECK(
+      K % splitk_factor == 0,
+      "splitk_factor ",
+      splitk_factor,
+      " must divide K ",
+      K,
+      " evenly for this benchmark");
+  int64_t Ki = K / splitk_factor;
+
+  auto options = at::TensorOptions().dtype(at::kHalf).device(at::kCUDA, 0);
+
+  // Architecture
+  auto properties = at::cuda::getDeviceProperties(0);
+  bool turing_or_later = properties->major >= 8 ||
+      (properties->major == 7 && properties->minor >= 5);
+
+  at::manual_seed(0);
+
+  auto fusion_ptr = std::make_unique<Fusion>();
+  auto fusion = fusion_ptr.get();
+  FusionGuard fg(fusion);
+
+  // Define fusion graph
+  auto a = makeContigTensor(3, DataType::Half);
+  auto b = makeContigTensor(3, DataType::Half);
+  fusion->addInput(a);
+  fusion->addInput(b);
+
+  at::Tensor aten_a, aten_b;
+  // Here we mimic splitting and transposing. For example, given an [M,K]
```

Comment:
Any reason why not using https://github.com/NVIDIA/Fuser/blob/a7df5a764db9dcf6ca8d29188fa06dde00a2c2a8/test/utils.h#L570C13-L570C36?

Reply:
Originally I did that to avoid alignment issues, since those methods place the batch dimension in the natural spot just like a split, unlike this code which allocates with the batch dimension first. When the M or N dimensions are not aligned we hit an error currently, so I was placing B first to make it easier to satisfy alignment. However, I think padding K is probably a better way to go, so the latest commit pads if necessary and prints a warning to stderr if so.